### PR TITLE
Implement sort most popular sessions view

### PR DESCRIPTION
**Pull request by odkhang**
This PR closes/references issue https://github.com/fossasia/eventyay-talk/issues/161 . It does so by:

Add sort dropdown in session view
implement logic when sort with title/time/popularity

## Summary by Sourcery

Implement a sorting feature in the calendar session view, allowing users to sort sessions by title, time, or popularity. Introduce dropdown components for filtering sessions by tracks, rooms, and session types, and enhance the UI for better responsiveness and usability.

New Features:
- Introduce a sorting feature in the calendar session view, allowing users to sort sessions by title, time, or popularity.
- Add dropdown components for filtering sessions by tracks, rooms, and session types.

Enhancements:
- Refactor the session filtering logic to support multiple criteria, including tracks, rooms, and session types.
- Improve the UI by adding a sticky position to the settings panel and adjusting the layout for better responsiveness.

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -4,31 +4,43 @@
...
-					path(d="m344.75 463.61h-117.92c-6.6289 0-11.84 5.2109-11.84 11.84s5.2109 11.84 11.84 11.84h117.92c5.2109 17.523 21.312 30.309 40.727 30.309 18.941 0 35.52-12.785 40.254-30.309h99.926c6.6289 0 11.84-5.2109 11.84-11.84s-5.2109-11.84-11.84-11.84h-99.926c-5.2109-17.523-21.312-30.309-40.254-30.309-19.418 0-35.52 12.785-40.727 30.309zm58.723 11.84c0 10.418-8.5234 18.469-18.469 18.469s-18.469-8.0508-18.469-18.469 8.5234-18.469 18.469-18.469 18.469 8.0508 18.469 18.469z")
-				template Filter
-				template(v-if="filteredTracks.length") ({{ filteredTracks.length }})
-			bunt-button.fav-toggle(v-if="favs.length", @click="onlyFavs = !onlyFavs; if (onlyFavs) resetFilteredTracks()", :class="onlyFavs ? ['active'] : []")
+			app-dropdown(v-for="item in filter")
```
[the remark this answers is not in the document]
Comment (hongquan) [reply src/App.vue:8]: Please update as Sourcery-AI suggested.
You will shoot yourself if you use `v-for` without proper `:key`.

File: src/components/LinearSchedule.vue [hunk 2]
```diff
@@ -46,52 +58,40 @@ export default {
...
+				const buckets = {}
+				for (const session of this.sessions) {
+					const key = session.start.format()
+					if (!buckets[key]) {
+						buckets[key] = []
+					}
+					if (!session.id) {
+						// Remove duplicate breaks, meaning same start, end and text
+						session.break_id = `${session.start}${session.end}${session.title}`
+						if (buckets[key].filter(s => s.break_id === session.break_id).length === 0) buckets[key].push(session)
```
Comment (hongquan) [inline src/components/LinearSchedule.vue:71]: Please use `Array.some` here, it is shorter than `Array.filter`.

File: src/components/LinearSchedule.vue [hunk 3]
```diff
@@ -46,52 +58,40 @@ export default {
...
-					buckets[key].push(session)
-				}
-			}
 
-			return Object.entries(buckets).map(([date, sessions]) => {
-				let sessionBucket = {}
+				return Object.entries(buckets).map(([date, sessions]) => ({
+					date: sessions[0].start,
+					// sort by room for stable sort across time buckets
+					sessions: sessions.sort((a, b) => this.rooms.findIndex(room => room.id === a.room.id) - this.rooms.findIndex(room => room.id === b.room.id))
```
Comment (hongquan) [inline src/components/LinearSchedule.vue:80]: This project already include lodash, please use "sort-by-key" strategy (https://lodash.com/docs/4.17.15#sortBy), it produces simpler code.

**Review by hongquan — CHANGES_REQUESTED** [review]

**Comment by odkhang** [comment]
Hi @hongquan, fixed as your suggestion, please help me to review it again

File: src/components/LinearSchedule.vue [hunk 4]
```diff
@@ -46,52 +59,40 @@ export default {
...
+				}))
+			} else {
+				const sortedSessions = this.sessions.slice().sort((a, b) => {
+					switch (this.sortBy) {
+						case 'title':
+							return a.title.localeCompare(b.title)
+						case 'popularity':
+							return b.fav_count - a.fav_count
+						default:
+							return _.sortBy(this.rooms, room => room.id === a.room.id) - _.sortBy(this.rooms, room => room.id === b.room.id)
```
Comment (hongquan) [inline src/components/LinearSchedule.vue:91]: Have you tested this code, how substraction between arrays work?

It seems that `array - array` always returns `NaN`.

**Review by hongquan — CHANGES_REQUESTED** [review]

File: src/components/LinearSchedule.vue (hunk 4 above)
Comment (odkhang) [reply src/components/LinearSchedule.vue]: Hi @hongquan,
you're right, array - array return NaN, 
currently we sort by time, title and popularity, so default case will not be reach, but I have removed it, default will sort by time. Please help me to review it again.

File: src/components/AppDropdownItem.vue [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    <div>
+        <slot/>
+    </div>
+</template>
+
+<script>
+export default {
+    name: "AppDropdownItem"
+};
+</script>
```
Comment (hongquan) [inline src/components/AppDropdownItem.vue:11]: Please add a blank line at the end of file and configure your editor to automatically do.

This rule is repeated again and again and you seem still don't configure your editor?

File: src/components/GridSchedule.vue [hunk 6]
```diff
@@ -338,7 +340,7 @@ export default {
 		min-width: min-content
 		> .room
 			position: sticky
-			top: calc(var(--pretalx-sticky-date-offset) + var(--pretalx-sticky-top-offset, 0px))
+			// top: calc(var(--pretalx-sticky-date-offset) + var(--pretalx-sticky-top-offset, 0px))
```
Comment (hongquan) [inline src/components/GridSchedule.vue:343]: Delete code, not comment out.

File: src/components/LinearSchedule.vue [hunk 7]
```diff
@@ -1,6 +1,6 @@
 <template lang="pug">
 .c-linear-schedule(v-scrollbar.y="")
-	.bucket(v-for="({date, sessions}, index) of sessionBuckets")
+	.bucket(v-if="sortBy == 'time'", v-for="({date, sessions}, index) of sessionBuckets")
```
Comment (hongquan) [inline src/components/LinearSchedule.vue:3]: There is comment about using `v-for` with `:key` but you still don't update?

**Review by hongquan — CHANGES_REQUESTED** [review]

File: src/components/LinearSchedule.vue (hunk 7 above)
Comment (odkhang) [reply src/components/LinearSchedule.vue]: Hi @hongquan, fixed as your suggestion, please help me review again

**Review by hongquan — APPROVED** [review]

**Pull request merged**